Harbor registry server scripts

sameerjvn opened the pull request:
This PR contains the following files 
- Harbor installer script and dependencies
- Upload and Download scripts for image push/pull. 
- List of dkube, datascience and other images used in Dkube 3.1.0.3
- Certs to enable HTTPS connection between docker and harbor and cert.sh to generate new ones.

deepio-oc reviewed (commented):
do we really need files of other repo bitnami-docker-harbor-portal-master checked into this repo ?

if we really need to use this files for bring up, lets add as a submodule. Did we make any changes in these files ?

deepio-oc commented on README.md:
```diff
@@ -1,2 +1,27 @@
-# registry-server
-Registry server for offline DKube install
+## To bring up Harbor Registry Server on Ubuntu :
```
Format of this file:

Description:

Registry Bringup
=============

Download Images
=============

Upload Images to Registry Server
========================

Update Docker Configuration on cluster nodes
==============================

sameerjvn replied:
@deepio-oc Addressed

deepio-oc commented on install.sh:
```diff
@@ -0,0 +1,17 @@
+sudo docker-compose up --detach
+sudo docker-compose logs -f -t > /tmp/harbor.log &
+echo "Triggered registry server bringup. Please wait..."
+./create_default_projects.sh $publicIP
```
go back to earlier directory at the end

sameerjvn replied:
This is not needed. On exiting install.sh, we are already in the same dir, not inside installer directory.

deepio-oc commented on install.sh:
```diff
@@ -0,0 +1,17 @@
+# Public IP of the registry VM
+publicIP="$(dig +short myip.opendns.com @resolver1.opendns.com)"
```
we can not use opendns since vm may not have internet access. Find a way to retrieve local ip. In case there are multiple local IPs, you can choose first one. Provide optional argument for user to specify the local IP.

We need to print the IP in the output "Using local IP x.x.x.x for registry-server".

sameerjvn replied:
Addressed.

deepio-oc commented on images/export.sh:
```diff
@@ -0,0 +1,35 @@
+                echo "No tag for $image"
+        fi
+done < $tmpfile
+rm $tmpfile
```
export.sh should generate a single image file containing

non-dkube images
dkube images (harcode the list of images which are not present in running cluster but used later on)
DS images (read from platform configmap)

sameerjvn replied:
@deepio-oc I have made it one file for dkube, non-dkube, datascience images.

deepio-oc commented on images/export.sh:
```diff
@@ -0,0 +1,35 @@
+# Run this script on a running k8s cluster to generate list of images used saved as ./images.txt
```
DKube cluster

sameerjvn replied:
Addressed.

deepio-oc commented on installer/docker-compose.yml:
```diff
@@ -0,0 +1,137 @@
+  core:
+    image: docker.io/bitnami/harbor-core:2.4.0
+    extra_hosts:
+      - "registry-server.dkube.io:34.71.3.5"
```
use a placeholder instead of IP

sameerjvn replied:
Addressed.

deepio-oc reviewed (commented):
1. remove server cert files from the folder since those certs are already present in the top folder.
2. remove server certs from certs folder and keep in installer/nginx/proxy folder.

sameerjvn commented:
@deepio-oc 
- I have shifted the `/etc/hosts` update logic to `update-dockerd.sh` since it contains the logic to ssh into all nodes.
- With this change, the user flow would be as follows:
1. `update-dockerd.sh`
2. `install.sh`
3. `download.sh` / `upload.sh`
- The flow is such because the upload script pushes to `registry-server.dkube.io` by default (instead of the harbor node IP) so it would be ideal if the domain name is already there in known hosts.
- Addressed above review comments
- Pending: README

sameerjvn commented:
@deepio-oc Addressed all review comments

deepio-oc commented on certs/registry-server.dkube.io.cert:
```diff
@@ -0,0 +1,34 @@
+-----BEGIN CERTIFICATE-----
```
can we delete this file?

sameerjvn replied:
This is needed for update-dockerd.sh

deepio-oc commented on certs/registry-server.dkube.io.csr:
```diff
@@ -0,0 +1,28 @@
+-----BEGIN CERTIFICATE REQUEST-----
```
can we delete this file?

deepio-oc commented on download.sh:
```diff
@@ -0,0 +1,67 @@
+done < "${list}"
+
+echo "Creating ${images} with $(echo ${pulled} | wc -w | tr -d '[:space:]') images"
+sudo docker save $(echo ${pulled})  > ${images}
```
This scripts needs to be tested.

deepio-oc commented on images/3.1.0.3.txt:
```diff
@@ -0,0 +1,195 @@
+rancher/mirrored-coredns-coredns:1.8.0
+rancher/mirrored-metrics-server:v0.4.1
+rancher/nginx-ingress-controller:nginx-0.43.0-rancher3
+rancher/rancher-agent:v2.5.9
```
rancher images should not be part of this. Customer would be installing rancher offline using ranching instructions which will download rancher images and upload to registry.

deepio-oc commented on images/export.sh:
```diff
@@ -0,0 +1,90 @@
+rm $tmpfile $imgsfile 2> /dev/null
+
+# Get all running images to get a list of non-dkube images
+kubectl get pods --all-namespaces -o jsonpath="{.items[*].spec.containers[*].image}" |\
```
use dkube namespaces here to limit images to just dkube images.

deepio-oc commented on install.sh:
```diff
@@ -0,0 +1,54 @@
+# Private IP of the registry VM
```
Local IP

deepio-oc commented on install.sh:
```diff
@@ -0,0 +1,54 @@
+# Private IP of the registry VM
+usage () {
+    echo "USAGE: $0 [--address x.x.x.x]"
+    echo "  [-i|--address IPAddres] Public/Private IP of the machine on which this script is to be run."
```
Local IP

sameerjvn commented:
- Fixes https://github.com/oneconvergence/gpuaas/issues/8757
- User has to have kubectl access on harbor node and run `./update-dockerd.sh --update-k8s-hosts` after kubernetes is installed.
- This will update the coredns configmap to map `registry-server.dkube.io` to the Harbor Node IP.

deepio-oc commented:
@sameerjvn updated dockerd script name suggest that we should be updating docker configuration but with the new option, it seems it will be doing completely different functionality.

we should either create another script (update-coredns.sh) or just mention the instructions to edit the configmap and manually insert the entry.

sameerjvn commented:
@deepio-oc Then I suppose we can add instructions for user to edit coredns configmap, as this script introduces kubectl access to harbor node as a requirement.
I shall add the lines in the README.
